src/EvalError.hs

```diff
@@ -0,0 +1,146 @@
+-- | Module EvalError defines errors that may occur during evaluation
+module EvalError
+  ( EvaluationError (..),
+    throwErr,
+  )
+where
+
+import Info
+import Obj
+import SymPath
+import TypeError
+import Util
+import Types
+
+-- | Errors that occur when evaluating forms.
+data EvaluationError
+  = SymbolNotFound SymPath
+  | PrivateBinding SymPath
+  | -- If
+    IfContainsNonBool XObj
+  | IfMalformed XObj
+  | -- Defn
+    DefnContainsQualifiedArgs XObj
+  | DefnIdentifierIsQualified XObj
+  | DefnContainsInvalidArgs XObj
+  | DefnMalformed XObj
+  | -- Def
+    DefIdentifierIsQualified XObj
+  | -- The
+    TheMalformed XObj
+  | -- Let
+    LetUnevenForms XObj
+  | LetMalformedIdentifiers [XObj]
+  | -- Fn
+    FnContainsQualifiedArgs XObj
+    -- With
+  |  WithContainsInvalidArgs XObj
+  | -- Do
+    DoMissingForms
+  | -- Unknown Form
+    UnknownForm XObj
+  | -- Function/Macro Application
+    MacroBadArgumentSplit [String]
+  | -- Address
+    AddressContainsNonSymbol XObj
+  | -- While
+    WhileContainsNonBool XObj
+  | -- Defmodule
+    ModuleRedefinition String
+  | DefmoduleContainsNonSymbol XObj
+  | DefmoduleNoArgs
+  | -- Load
+    LoadInvalidArgs [XObj]
```

Comment:
While it’s certainly more type safe this way, wouldn’ it increase reusability if we had a single `InvalidArgs` error?

Reply:
Good point! I think that'd be more consistent with primitive error handling too, right?

I do sort of like having errors specialized to each type of form, but I don't have a strong preference either way, so I'm happy to change it. Less code is better code, so let's try it the way you suggested!

Reply:
@hellerve Can you please explain the issue a little bit more, I'm not sure I follow.

Reply:
At the moment, there is a `LoadInvalidArgs`, a `LoadOnceInvalidArgs`, a `SetInvalidArgs`, a `WithContainsInvalidArgs`, and a `DefnContainsInvalidArgs`. Since the error message is mostly the same, I would argue we should make a `InvalidArgs String XObj` instead, with the first argument being the name of the form that got invalid arguments (could also be an `XObj` with a `Sym`, of course).

Does that make sense?